src/transformers/models/superglue/modeling_superglue.py

```diff
@@ -0,0 +1,883 @@
+                Index of keypoint matched in the second image.
+            matches_1 (`torch.Tensor` of shape `(1, num_keypoints)`):
+                Index of keypoint matched in the first image.
+            matching_scores_0 (`torch.Tensor` of shape `(1, num_keypoints)`):
+                Scores of predicted matches from the first image.
+            matching_scores_1 (`torch.Tensor` of shape `(1, num_keypoints)`):
+                Scores of predicted matches from the second image.
+            all_hidden_states (`tuple(torch.FloatTensor)`, *optional*):
+                Tuple of `torch.FloatTensor` (one for the output of each stage) of shape `(1, 2, num_keypoints,
+                num_channels)`.
+            all_attentions (`tuple(torch.FloatTensor)`, *optional*):
+                Tuple of `torch.FloatTensor` (one for each layer) of shape `(1, 2, num_heads, num_keypoints,
+                num_keypoints)`.
+        """
+        all_hidden_states = () if output_hidden_states else None
+        all_attentions = () if output_attentions else None
+
+        if image0_keypoints.shape[1] == 0 or image1_keypoints.shape[1] == 0:  # no keypoints
+            shape0, shape1 = image0_keypoints.shape[:-1], image1_keypoints.shape[:-1]
+            return (
+                image0_keypoints.new_full(shape0, -1, dtype=torch.int),
+                image1_keypoints.new_full(shape1, -1, dtype=torch.int),
+                image0_keypoints.new_zeros(shape0),
+                image1_keypoints.new_zeros(shape1),
+                all_hidden_states,
+                all_attentions,
+            )
+
+        descriptors_0 = torch.transpose(image0_descriptors, -1, -2)
+        descriptors_1 = torch.transpose(image1_descriptors, -1, -2)
+
+        # Keypoint normalization
+        keypoints0 = normalize_keypoints(image0_keypoints, height, width)
+        keypoints1 = normalize_keypoints(image1_keypoints, height, width)
+
+        encoded_keypoints0 = self.keypoint_encoder(
+            keypoints0, image0_scores, output_hidden_states=output_hidden_states
+        )
+        encoded_keypoints1 = self.keypoint_encoder(
+            keypoints1, image1_scores, output_hidden_states=output_hidden_states
+        )
+
+        last_hidden_state_0 = encoded_keypoints0[0]
+        last_hidden_state_1 = encoded_keypoints1[0]
+
+        # Keypoint MLP encoder.
+        descriptors_0 = descriptors_0 + last_hidden_state_0
+        descriptors_1 = descriptors_1 + last_hidden_state_1
+
+        # Multi-layer Transformer network.
+        gnn_outputs = self.gnn(
+            descriptors_0,
+            descriptors_1,
+            output_hidden_states=output_hidden_states,
+            output_attentions=output_attentions,
+        )
+        descriptors_0, descriptors_1 = gnn_outputs[:2]
+
+        # Final MLP projection.
+        projected_descriptors_0 = self.final_projection(descriptors_0)
+        projected_descriptors_1 = self.final_projection(descriptors_1)
```

Comment:
Given the projection layer here - we could split this module up into `SuperGlueModel` which contains everything up to the outputs of the GNN, and then SuperGlueForKeypointMatching, which contains the super glue model and the final projection layer

Reply:
Is this required out of convention in transformers ? What is the motivation behind having the output of the GNN ?